### ISSUE TITLE
Can't aggregate `connectionDetails` of multiple managed resources into secret of composite.

### ISSUE THREAD

**Issue by jonnylangefeld** [crossplane/crossplane#3890]
I discovered this issue with more complex composite resources, but I've reproduced the issue to a GCP service account and service account key resource to make the problem easier to understand.

I have the following `XRD`, `Composition` and `Claim` resources:

<details>
<summary>`XRD` yaml:</summary>

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: CompositeResourceDefinition
metadata:
  name: xcruiseserviceaccounts.provisioning.getcruise.com
spec:
  group: provisioning.getcruise.com
  names:
    kind: XCruiseServiceAccount
    plural: xcruiseserviceaccounts
  claimNames:
    kind: CruiseServiceAccount
    plural: cruiseserviceaccounts
  connectionSecretKeys:
    - key.json
  versions:
    - name: v1alpha1
      served: true
      referenceable: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec: {}
```
</details>

<details>
<summary>`Composition` yaml:</summary>

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: xcruiseserviceaccounts.provisioning.getcruise.com
  labels:
    provider: gcp
spec:
  publishConnectionDetailsWithStoreConfigRef:
    name: default
  compositeTypeRef:
    apiVersion: provisioning.getcruise.com/v1alpha1
    kind: XCruiseServiceAccount
  resources:
    - name: serviceaccount
      base:
        # https://doc.crds.dev/github.com/upbound/provider-gcp/cloudplatform.gcp.upbound.io/ServiceAccount/v1beta1
        apiVersion: cloudplatform.gcp.upbound.io/v1beta1
        kind: ServiceAccount
        spec:
          forProvider: {}
      patches:
        - type: CombineFromComposite
          toFieldPath: spec.forProvider.description
          policy:
            fromFieldPath: Required
          combine:
            variables:
              - fromFieldPath: spec.claimRef.namespace
              - fromFieldPath: spec.claimRef.name
            strategy: string
            string:
              fmt: "%s/%s"
        - type: ToCompositeFieldPath
          fromFieldPath: status.atProvider.email
          toFieldPath: status.serviceAccountEmail
          policy:
            fromFieldPath: Required
    - name: serviceaccountkey
      base:
        # https://doc.crds.dev/github.com/upbound/provider-gcp/cloudplatform.gcp.upbound.io/ServiceAccountKey/v1beta1
        apiVersion: cloudplatform.gcp.upbound.io/v1beta1
        kind: ServiceAccountKey
        spec:
          forProvider:
            publicKeyType: TYPE_X509_PEM_FILE
            serviceAccountIdSelector:
              matchControllerRef: true
          publishConnectionDetailsTo:
            configRef:
              name: default
      patches:
        - type: CombineFromComposite
          combine:
            variables:
              - fromFieldPath: metadata.uid
            strategy: string
            string:
              fmt: "%s-serviceaccountkey"
          toFieldPath: spec.publishConnectionDetailsTo.name
          policy:
            fromFieldPath: Required
      connectionDetails:
        - fromConnectionSecretKey: private_key
          name: key.json
```
</details>

<details>
<summary>`Claim` yaml:</summary>

```yaml
apiVersion: provisioning.getcruise.com/v1alpha1
kind: CruiseServiceAccount
metadata:
  name: demo-sa
spec: {}
```
</details>

So far so good, this gives me the following Kubernetes secrets, as expected:
```
ce675139-8fa7-42b9-9c9e-0aab8cd7daf5                       connection.crossplane.io/v1alpha1     1      34m
ce675139-8fa7-42b9-9c9e-0aab8cd7daf5-serviceaccountkey     connection.crossplane.io/v1alpha1     2      35m
```

The 2 secret values in the `managed` resource secret (as indicated by the `-serviceaccountkey` suffix) are `private_key` and `attribute.private_key`, which are the same value (which is confusing, but I guess that's an implementation detail of the provider and not the point of this issue)

The `private_key` secret successfully gets propagated into the `composite` secret, which just has the uid as name. In there it's called `key.json` as instructed by the
```
      connectionDetails:
        - fromConnectionSecretKey: private_key
          name: key.json
```
part.

### The issue
Now if you add a second `ServiceAccountKey` to the composition like so:
```
    - name: serviceaccountkey1
      base:
        # https://doc.crds.dev/github.com/upbound/provider-gcp/cloudplatform.gcp.upbound.io/ServiceAccountKey/v1beta1
        apiVersion: cloudplatform.gcp.upbound.io/v1beta1
        kind: ServiceAccountKey
        spec:
          forProvider:
            publicKeyType: TYPE_X509_PEM_FILE
            serviceAccountIdSelector:
              matchControllerRef: true
          publishConnectionDetailsTo:
            configRef:
              name: default
      patches:
        - type: CombineFromComposite
          combine:
            variables:
              - fromFieldPath: metadata.uid
            strategy: string
            string:
              fmt: "%s-serviceaccountkey1"
          toFieldPath: spec.publishConnectionDetailsTo.name
          policy:
            fromFieldPath: Required
      connectionDetails:
        - fromConnectionSecretKey: private_key
          name: key1.json
```

I would expect one additional secret with suffix `-serviceaccountkey1` and the composite secret with name `ce675139-8fa7-42b9-9c9e-0aab8cd7daf5` to have two keys, `key.json` and `key1.json`. But instead the composite secret still only has one key, `key.json`:

```
ce675139-8fa7-42b9-9c9e-0aab8cd7daf5                       connection.crossplane.io/v1alpha1     1      34m
ce675139-8fa7-42b9-9c9e-0aab8cd7daf5-serviceaccountkey     connection.crossplane.io/v1alpha1     2      35m
ce675139-8fa7-42b9-9c9e-0aab8cd7daf5-serviceaccountkey1    connection.crossplane.io/v1alpha1     2      35m
```

```
╰─ kg secret -n crossplane ce675139-8fa7-42b9-9c9e-0aab8cd7daf5 -o yaml
apiVersion: v1
data:
  key.json: <base64 of key>
kind: Secret
metadata:
  creationTimestamp: "2023-03-22T00:16:13Z"
  labels:
    secret.crossplane.io/owner-uid: ce675139-8fa7-42b9-9c9e-0aab8cd7daf5
  name: ce675139-8fa7-42b9-9c9e-0aab8cd7daf5
  namespace: crossplane
  resourceVersion: "1467750019"
  uid: a7ed013c-6fb5-4630-a422-1af9fb3e8ac5
type: connection.crossplane.io/v1alpha1
```

It should have two keys, `key.json` and `key1.json`.

### How can this be improved?

It's not an uncommon use case to have multiple managed resources in a composition that all expose connection details. And we would like to aggregate these connection details into the secret of the composite resource and further obviously into the secret exposed to the user of the `claim`. But that doesn't seem possible currently.

### Things I've tried:

* Sometimes when you update an existing composition and post it to the cluster, not all resources get updated. I excluded this as being the root cause by creating an entire new claim after the composition was already updated with the two service account keys. The issue remained
* The `connectionDetails` allow for a `type: FromConnectionSecretKey` setting. I don't think that should influence it as I'm already using `fromConnectionSecretKey: private_key`. But just in case I set `type: FromConnectionSecretKey` on both service account keys and it made no difference.
* On the XRD you set the list of `connectionSecretKeys`. I'm thinking that it shouldn't matter if that list includes `key1.json`, as that's only the secret that gets exposed to the user. Even if that list only contains `key.json`, I'd expect the composite secret to have both secrets. But just to exclude this as being the root cause I added both, `key.json` and `key1.json` to that list and it made no difference

### What environment did it happen in?
Crossplane version: `v1.10.1`
GCP provider version: `v0.27.0`

This setup doesn't include any external secret stores, the `configRef.name: default` is just the regular Kubernetes secret store.

If this is just a misconfiguration on my part I'm happy to close this issue again. But I verified with co-workers and we all think the API fields we specified look all right.

**Comment by jonnylangefeld** [comment]
Furthermore, if you delete the first `ServiceAccountKey` resource from the list of resources in the composition and post the composition to the cluster, then one of the `ServiceAccountKey` resources disappears from the cluster and also the `ce675139-8fa7-42b9-9c9e-0aab8cd7daf5-serviceaccountkey` secret (the one without the `1` at the end) disappear, as expected.

However, the composite secret with name `ce675139-8fa7-42b9-9c9e-0aab8cd7daf5` still only has the `key.json`, the connection detail from the just removed resource. So now these two secrets exist, but the private key value in them doesn't match: 
```
ce675139-8fa7-42b9-9c9e-0aab8cd7daf5                       connection.crossplane.io/v1alpha1     1      3h44m
ce675139-8fa7-42b9-9c9e-0aab8cd7daf5-serviceaccountkey1    connection.crossplane.io/v1alpha1     2      3h44m
```

After reconciliation I would expect the `key.json` key to be removed from the composite secret and `key1.json` to be added.

**Comment by jonnylangefeld** [comment]
Sorry for multiple posts in a row, but as I keep trying out more things I make new findings. I didn't mean this to turn into a stream of consciousness. But this should be the last post.

I now think this issue is closely related to #2024, although not obvious at first sight. #2024 describes (and I ran into this myself in #3824 as well) how you can't update the secret of an existing resource. In this issue report I'm talking about how you also cannot add an entire new resource that has secrets itself.

Reading through @muvaf's [comment here](https://github.com/crossplane/crossplane/issues/2024#issuecomment-744524336), it sounds like it's all related to the list of `connectionSecretKeys` on the XRD, which apparently only gets read upon creation of the XRD. That means that this design doesn't only affect if you update the `connectionDetails` of one of the resources in the composition, but also if you add an entire new resource to the list in the composition.

I verified this by deleting the claim, composition and xrd and deployed it again with just the `serviceaccountkey1` resource. After re-applying my claim, sure enough I got a composite secret with the `key1.json` entry as expected.
Lastly, I tried to remove the claim, composition and xrd again, added both `ServiceAccountKeys` as in my initial test and applied it all again, and then I got the desired composite secret with the keys `key.json` and `key1.json`.

So I believe a fix for #2024 will fix this behavior as well. So feel free to close this issue if you think it's the same root cause. We can see this issue as yet another indicator of importance to fix #2024. We will definitely have the use case to eventually add resources with secrets to already existing compositions and we cannot delete all claims in order to delete all XRDs as many of these resources will be data stores of various users.
